src/components/pages/lessons/overlay/confirm-membership.tsx

```diff
@@ -250,15 +252,29 @@ const ConfirmMembership: React.FC<ConfirmMembershipProps> = ({
   if (loadingSession || currentState.matches('pending'))
     return <LoadingSession />
 
+  const {session_id, ...nonSessionQueryParams} = router.query
+
+  const cleanUrlAndViewLesson = () => {
+    router.replace({
+      pathname: router.pathname,
+      query: nonSessionQueryParams,
+    })
+
+    setTimeout(viewLesson, 750)
+  }
```

Comment:
This removes `session_id` query param from the URL and then triggers `viewLesson`. In my testing if `viewLesson` was called immediately after `router.replace`, the updated URL hadn't propagated yet causing the machine state to not update properly. This `timeout` gives it a chance to propagate.

This is a reasonable near-term solution, but it is admittedly a bit hacky. So we can keep this in mind and, at some point, look at more involved alternatives like moving this logic into the machine.